src/process_utility.c

```diff
@@ -1200,18 +1204,110 @@ static void
 process_index_chunk(Hypertable *ht, Oid chunk_relid, void *arg)
 {
 	CreateIndexInfo *info = (CreateIndexInfo *) arg;
+
 	IndexStmt  *stmt = transformIndexStmt(chunk_relid, info->stmt, NULL);
 	Chunk	   *chunk = ts_chunk_get_by_relid(chunk_relid, ht->space->num_dimensions, true);
 
 	ts_chunk_index_create_from_stmt(stmt, chunk->fd.id, chunk_relid, ht->fd.id, info->obj.objectId);
 }
 
+static void
+process_index_chunk_multitransaction(Hypertable *ht, Oid chunk_relid, void *arg)
+{
+	CreateIndexInfo *info = (CreateIndexInfo *) arg;
+	Chunk	   *chunk;
+	Relation	hypertable_rel;
+	Relation	hypertable_index_rel;
+	Relation	chunk_rel;
+	IndexStmt  *stmt;
+	CatalogSecurityContext sec_ctx;
+
+	/*
+	 * commit root table's transaction if it exists (foreach_chunk requires a
+	 * transcation running, so we cannot commit before this point)
+	 */
+	if (ActiveSnapshotSet())
+		PopActiveSnapshot();
+
+	if (IsTransactionOrTransactionBlock())
+		CommitTransactionCommand();
+
+	/* Start a new transaction for each relation. */
+	StartTransactionCommand();
+	/* functions in indexes may want a snapshot set */
+	Snapshot	snapshot = GetTransactionSnapshot();
+
+	PushActiveSnapshot(snapshot);
+
+	/*
+	 * Change user since chunk's are typically located in an internal schema
+	 * and chunk indexes require metadata changes. In regular  CREATE INDEX we
+	 * do this in process_index_end
+	 */
+	ts_catalog_database_info_become_owner(ts_catalog_database_info_get(), &sec_ctx);
+
+	/*
+	 * Hold a lock on the hypertable, it's index index, and the chunk to
+	 * prevent from being altered. Since we use the same relids across
+	 * transactions, there is a potential issue if the id gets reassigned
+	 * between one sub-transaction and the next. CLUSTER has a similar issue.
+	 */
+	hypertable_rel = relation_open(ht->main_table_relid, AccessShareLock);
+	chunk_rel = relation_open(chunk_relid, AccessShareLock);
+	hypertable_index_rel = relation_open(info->obj.objectId, AccessShareLock);
+
+	stmt = transformIndexStmt(chunk_relid, info->stmt, NULL);
+
+	chunk = ts_chunk_get_by_relid(chunk_relid, ht->space->num_dimensions, true);
+
+	/*
+	 * use ts_chunk_index_create instead of ts_chunk_index_create_from_stmt to
+	 * handle cases where the index is altered
+	 */
+	ts_chunk_index_create(hypertable_rel, ht->fd.id, hypertable_index_rel, chunk->fd.id, chunk_rel);
+
+	relation_close(hypertable_index_rel, NoLock);
```

Comment:
Shouldn't you release the lock? (the same question for two lines below)

Reply:
Running relation_close with NoLock means that the lock is released.

Reply:
Hmmm...are you sure about that? Code comment for `relation_close` is telling:
`If lockmode is not "NoLock", we then release the specified lock.`. 
Also implementation of `relation_close`:
```
if (lockmode != NoLock)
	UnlockRelationId(&relid, lockmode);
```

Reply:
@niksajakovljevic The lock will be closed when the transaction commits.